Add model-centric UI overview page

edmundito opened the pull request:
## Change description

This adds the first pass of the model-centric overview page:

`/models/:modelId/overview`

![image](https://user-images.githubusercontent.com/168664/144083747-e064fcb2-78a2-4a68-a312-6305b0e7c0b8.png)

![image](https://user-images.githubusercontent.com/168664/144083812-b45bf6fc-63cd-4809-902e-cd0f17dce3de.png)


https://user-images.githubusercontent.com/168664/144083705-44088d67-c84c-4712-b61f-3f31526ca389.mov



Includes:
- Show all data
- Link to all data
- Actions for data
- Collapsible sections

To be added separately:
- Single API call to request all data
- Updated navigation to point to this page

## Checklists

### Development

- [x] Application changes have been tested appropriately

### Impact

- [ ] Code follows company security practices and guidelines
- [ ] Security impact of change has been considered
- [x] Performance impact of change has been considered
- [ ] Possible migration needs considered (model migrations, config migrations, etc.)

Please explain any security, performance, migration, or other impacts if relevant:

Performance:  The page is currently making multiple requests to get all the data. We will consider consolidating all the calls into a single one

### Code review

- [x] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached where applicable.
- [x] Relevant tags have been added to the PR (bug, enhancement, internal, etc.)

edmundito committed:
Update useAPI hook and client with better TS typing

edmundito committed:
Add first pass unified model overview page

* Update sources to use common formatSchedule util shared between source table and overview page

edmundito committed:
Reorganize model overview and add basic group rendering
* UpdatePageHeader prop defs

edmundito committed:
Add schedules list to model overview
* Add sync-alt icon to Icons component

edmundito committed:
Ensure items have correct title and status
* Add formatName util
* Ensure other pages use formatName util

edmundito committed:
Rename model overview components

edmundito committed:
Move model overview components into their own files

edmundito committed:
Add link to collection tracked in model overview
* Extract DestinationCollectionLink into its own component

edmundito committed:
Update model overview components from Summary to Info

edmundito committed:
Update model overview with links to sources, groups, schedules, and destinations

edmundito committed:
EntityInfoHeading -> EntityInfoHeader

edmundito committed:
Update all api calls to model overview page to be requested in parallel

edmundito committed:
Add buttons to model-centric ui
* Add grouparooModelContext to provide grouparoo model to different components without having to pass down the context

edmundito committed:
Update model overview schedules to run schedules

edmundito committed:
Fix grouparooModelContext import case

edmundito committed:
Use RunAllSchedulesButton in model sources page

edmundito committed:
Move ModelOverviewCard to own component file

edmundito committed:
Move add destination button to top right
* Update ModelOverviewCard to accept action components to render right of heading

edmundito committed:
Ensure all action buttons are alighed to the top right of their container

* Update SectionContainer to support action buttons
* Update ModelOverviewCard header alighment to better fit buttons and heading

edmundito commented on ui/ui-components/components/model/overview/EntityInfoContainer.tsx:
```diff
@@ -0,0 +1,20 @@
+  );
+};
+
+export default EntityInfoContainer;
```
Still working on this name. An "entity" is a a Grouparoo object such as a source, destination, group, or grouparoo model. I could also call it a "GrouparooObjectContainer". I didn't want to call it model because of the already overloaded use of the term.

evantahler replied:
I don't hate the name!  Naming is hard.

edmundito commented on ui/ui-components/components/schedule/RunAllSchedulesButton.tsx:
```diff
@@ -0,0 +1,59 @@
+  );
+};
+
+export default RunAllSchedulesButton;
```
This introduces a new pattern into the ui-components of what I call "smart components" in that they also contain their own logic allowing things like this button to be shared in multiple pages (sources list and the model overview)

krishnaglick replied:
We should avoid default exports outside of Pages as it makes refactor actions far more difficult.

edmundito commented on ui/ui-components/pages/model/[modelId]/overview.tsx:
```diff
@@ -0,0 +1,168 @@
+  );
+};
+
+export const getServerSideProps: GetServerSideProps<Props> = async (
```
This uses `getServerSideProps` over `getInitialProps` to move away from deprecated features. However, you can see that in ui-enterprise the import gets a little more complex

https://nextjs.org/docs/api-reference/data-fetching/getInitialProps

edmundito commented on ui/ui-components/pages/model/[modelId]/overview.tsx:
```diff
@@ -0,0 +1,168 @@
+    modelId,
+  };
+
+  // TODO: Create a single api call that requests all data for the model overview page.
```
Unless the call below is good enough? One of the things I haven't thought about is the limit/offset and how that should work for all the objects.

evantahler replied:
from standup - for now we don't need to make a single API call

edmundito committed:
grouparooModel -> grouparooModelContext

edmundito committed:
Add model overview page to ui-config and ui-community

edmundito commented on ui/ui-components/contexts/grouparooModel.tsx:
```diff
@@ -0,0 +1,15 @@
+  <GrouparooModelContext.Provider value={model}>
+    {children}
+  </GrouparooModelContext.Provider>
+);
```
This is another new pattern I'm introducing. This sets up everything that is needed for a context such that the context user just needs to either use the rich `GrouparooModelContextProvider`provider or the rich `useGrouparooModelContext` hook. This is pretty simple but as more logic may be added the provider can be extended to manage the value of the provider.

krishnaglick replied:
Cloud is doing something similar! 👍

evantahler replied:
👍 

I imagine one day contexts replace the `handlers` we have in the UI

krishnaglick commented on ui/ui-components/components/PageHeader.tsx:
```diff
@@ -4,21 +4,23 @@ import ModelIcon from "./ModelIcon";
+interface Props {
+  title: React.ReactNode;
+  icon?: string;
+  badges?: any[];
```
I feel like this could be typed?

krishnaglick commented on ui/ui-components/components/model/overview/EntityInfoContainer.tsx:
```diff
@@ -0,0 +1,20 @@
+}
+
+const EntityInfoContainer: React.FC<Props> = ({ app, children }) => {
+  return app && app.icon ? (
```
```suggestion
  return app?.icon ? (
```

krishnaglick commented on ui/ui-components/components/model/overview/EntityInfoHeader.tsx:
```diff
@@ -0,0 +1,35 @@
+import { formatName } from "../../../utils/formatName";
+import StateBadge from "../../badges/StateBadge";
+
+type Entity = Parameters<typeof formatName>[0];
```
Why not just export this type from `formatName`?

edmundito replied:
Fixed

edmundito committed:
Add empty and disabled state

krishnaglick commented on ui/ui-components/components/model/overview/ModelOverviewDestinations.tsx:
```diff
@@ -0,0 +1,74 @@
+  destination,
+}) => {
+  const { app, connection, exportTotals } = destination;
+  const isCommunityUI = grouparooUiEdition() === "community";
```
Could be hoisted outside of the component.

edmundito replied:
I'm not sure what this function does, wether it's resolved at runtime or not

evantahler replied:
It's resolved at runtime via ENV variables

krishnaglick replied:
@evantahler This data is available immediately to the app, no?

krishnaglick commented on ui/ui-components/components/model/overview/ModelOverviewDestinations.tsx:
```diff
@@ -0,0 +1,74 @@
+  );
+};
+
+interface Props {
```
```suggestion
interface ModelOverviewDestinationsProps {
```

edmundito replied:
This is preferential as unless we need export these props

krishnaglick replied:
I'll bow to convention

krishnaglick commented on ui/ui-components/components/model/overview/ModelOverviewDestinations.tsx:
```diff
@@ -0,0 +1,74 @@
+  destinations?: Models.DestinationType[];
+}
+
+const ModelOverviewDestinations: React.FC<Props> = ({ destinations }) => {
```
```suggestion
const ModelOverviewDestinations: React.FC<ModelOverviewDestinationsProps> = ({ destinations }) => {
```
Not that this specific name needs to be used, but I find it's good to name props.

krishnaglick commented on ui/ui-components/components/model/overview/ModelOverviewPrimarySource.tsx:
```diff
@@ -0,0 +1,42 @@
+    <SectionContainer
+      title="Primary Source"
+      iconType={source ? "app" : "icon"}
+      icon={source ? source.app?.icon : "file-import"}
```
```suggestion
      icon={source?.app?.icon ?? "file-import"}
```

krishnaglick commented on ui/ui-components/components/model/overview/SectionContainer.tsx:
```diff
@@ -0,0 +1,49 @@
+}) => {
+  return (
+    <Media>
+      {icon && iconType === "app" ? (
```
I think if you add `typeof icon === "string` in here you won't need the `as` below.

edmundito replied:
Fancy

krishnaglick commented on ui/ui-components/pages/model/[modelId]/overview.tsx:
```diff
@@ -0,0 +1,144 @@
+  destinations?: Models.DestinationType[];
+}
+
+const Page: NextPage<Props & { ctx: any; errorHandler: any }> = ({
```
This could be more precisely named, maybe `ModelPage`?

edmundito replied:
This is the general naming convention for all pages in the project

krishnaglick replied:
Ah, okay.

edmundito committed:
Add EntityType to models object and update references

edmundito committed:
Update overview copy to be more clear, capitalize grouparoo objects

edmundito committed:
Rebuild PageHeader away from table elements

edmundito committed:
Add actions to PageHeader and add edit action to model overview page header

edmundito committed:
Update right margin for PageHeader icon

edmundito committed:
Code cleanup

edmundito commented:
I thought I was done with the code changes but I added the empty / disabled states, updated the copy, and forgot about the edit model icon earlier so now it's there.

edmundito committed:
Update SectionContainer and EntityInfoContainer app icon size to match non-app icons

edmundito committed:
Format primary key mapping in SourceInfo component

edmundito committed:
Improve schedule formatting

edmundito committed:
CAdd collapsible sections to model overview

edmundito committed:
GrouparooEntityTypes -> EntityTypes

edmundito commented:
I had some extra time to add the collapsible sections so I pushed the change. I updated the screenshots and video in the description to reflect the current version.

edmundito committed:
Remove unused fragment from EntityList

edmundito committed:
Fix expanded setter name in EntityList

evantahler commented on ui/ui-components/client/client.ts:
```diff
@@ -108,16 +115,16 @@ export class Client {
-    res?
-  ) {
-    const headers = {
+    req?: IncomingMessage,
+    res?: ServerResponse
```
👍 nice!

evantahler commented on ui/ui-components/components/schedule/RunAllSchedulesButton.tsx:
```diff
@@ -0,0 +1,73 @@
+  }
+
+  async function enqueueAllSchedulesRun() {
+    onStart?.();
```
this is a cool pattern!

krishnaglick commented on ui/ui-components/components/PageHeader.tsx:
```diff
@@ -1,75 +1,99 @@
+  );
+
+  const badgesNode: React.ReactNode =
+    badges && !!badges.length
```
```suggestion
    badges?.length
```

krishnaglick commented on ui/ui-components/components/PageHeader.tsx:
```diff
@@ -1,75 +1,99 @@
+      : null;
+
+  const actionsNode: React.ReactNode =
+    actions && !!actions.length
```
```suggestion
    actions?.length
```

krishnaglick commented on ui/ui-components/components/PageHeader.tsx:
```diff
@@ -1,75 +1,99 @@
+  const actionsNode: React.ReactNode =
+    actions && !!actions.length
+      ? actions.map((action, idx) => (
+          <Fragment key={`header_action_${idx}`}>{action} </Fragment>
```
Does `action` have an id or something on it? React recommended against using an index as part of the key iirc.

krishnaglick commented on ui/ui-components/components/PageHeader.tsx:
```diff
@@ -1,75 +1,99 @@
-}
+};
+
+export default PageHeader;
```
We should explicitly export this, not use a default.
https://humanwhocodes.com/blog/2019/01/stop-using-default-exports-javascript-module/
A decent article on it.

edmundito replied:
Let's follow up about this separately because is how it's currently done in all of ui-components

krishnaglick commented on ui/ui-components/components/model/overview/EntityList.tsx:
```diff
@@ -0,0 +1,96 @@
+  return items.map((item, index) => {
+    const itemId =
+      itemType === "schedule"
+        ? (item as Models.ScheduleType).sourceId
```
Could be a good use-case for a Typeguard, but it's just a slightly more elegant version of this.

krishnaglick commented on ui/ui-components/components/model/overview/ModelOverviewDestinations.tsx:
```diff
@@ -0,0 +1,89 @@
+      }
+    >
+      <ListGroup className="list-group-flush">
+        {destinations && !!destinations.length ? (
```
```suggestion
        {destinations?.length ? (
```

krishnaglick commented on ui/ui-components/components/schedule/RunAllSchedulesButton.tsx:
```diff
@@ -0,0 +1,73 @@
+      setIsLoading(false);
+      onComplete?.();
+    }
+  }
```
This may benefit from being wrapped in a `useCallback`

evantahler reviewed (approved):
Lots of great refactoring in here!

I'm going to start a document where we can discuss our font-end coding conventions and how we can (slowly) work on updating the rest of the front-end codebase

krishnaglick commented on ui/ui-components/pages/model/[modelId]/overview.tsx:
```diff
@@ -0,0 +1,158 @@
+  const sources = useMemo(() => {
+    const result = [...secondarySources];
+    if (primarySource) {
+      result.splice(0, 0, primarySource);
```
```suggestion
      result.unshift(primarySource);
```

edmundito committed:
Move shared model overview components to lib subfolder and add SeparatedItems component

edmundito committed:
Update actions prop in ManagedCard and SectionContainer to use array of nodes

edmundito commented:
I addressed the comment and did some cleanup of the common components.

edmundito committed:
Make managed card actions optional

edmundito committed:
Fix SeparatedItems separator default value

edmundito committed:
Syntax cleanup
